### PR TITLE
Added DuckDuckGo support and create a common interface for search engines

### PR DESCRIPTION
**Pull request by Vishnuxx**
This PR introduces DuckDuckGo support and refactored the code in `src/browser/search.py` file to create a common interface for accessing search engines.

### Changes Made:
* Added `duckduckgo_search` dependency in requirements.txt

* `src/browser/search_engines/Bing.py` : Contains the `BingSearch `class for Bing search functionality.

* `src/browser/search_engines/DuckDuckGo.py` : Contains the `DuckDuckGoSearch` class for DuckDuckGo search functionality.

* Removed previous `BingSearch` implementation from `src/browser/search.py`.
* Introduced `WebSearch` class into `src/browser/search.py` as the common interface for web search.

### CODE REVIEW

**Comment by Vishnuxx** [comment]
This PR fixes these issues : #186 #97 #65 #38

**Comment by mufeedvh** [comment]
Hey @Vishnuxx, thanks for the PR! Looks good to me, would you be able to resolve the merge conflicts? Will merge this right away! Also, this collides with #99 and #70 right? Can you please take a look into that as well? 🙌

**Comment by Vishnuxx** [comment]
Hello @mufeedvh, thank you for your response! The conflicts have been resolved, and this pull request is now ready for merging. Please review once again and proceed with the merge.

File: src/agents/researcher/researcher.py [hunk 1]
```diff
@@ -4,13 +4,13 @@
...
 from src.llm import LLM
-from src.browser.search import BingSearch
+from src.browser.search import WebSearch
 
 PROMPT = open("src/agents/researcher/prompt.jinja2").read().strip()
 
 class Researcher:
     def __init__(self, base_model: str):
-        self.bing_search = BingSearch()
+        self.bing_search = WebSearch()
```
Comment (Tineshnehete) [inline src/agents/researcher/researcher.py:13]: As we are using a common class for search engine we can rename ```bing_search``` to ```web_search```

File: src/agents/agent.py [hunk 2]
```diff
@@ -61,14 +62,14 @@ def __init__(self, base_model: str):
 
     def search_queries(self, queries: list, project_name: str) -> dict:
         results = {}
-        
+
         knowledge_base = KnowledgeBase()
-        bing_search = BingSearch()
+        bing_search = WebSearch()
```
Comment (Tineshnehete) [inline src/agents/agent.py:67]: As we are using a common class for search engine we can rename bing_search to web_search

**Comment by Vishnuxx** [comment]
Hii @Tineshnehete , thanks for reviewing the PR!

File: src/browser/search.py [hunk 3]
```diff
@@ -1,25 +1,25 @@
...
+    DUCKDUCKGO = DuckDuckGoSearch()
+
+
+class WebSearch:
     def __init__(self):
-        self.config = Config()
-        self.bing_api_key = self.config.get_bing_api_key()
-        self.bing_api_endpoint = self.config.get_bing_api_endpoint()
+        # self.config = Config()
+        self.search_engine = SearchEngine.DUCKDUCKGO.value
```
Comment (Tineshnehete) [inline src/browser/search.py:14]: Couldn't We make to enable user to set search engine class via config ?
```suggestion
        self.search_engine = load_class(self.config.get_search_engine_class()) // replace load class method
```

File: src/browser/search.py [hunk 4]
```diff
@@ -1,25 +1,25 @@
...
+    BING = BingSearch()
+    DUCKDUCKGO = DuckDuckGoSearch()
+
+
+class WebSearch:
     def __init__(self):
-        self.config = Config()
-        self.bing_api_key = self.config.get_bing_api_key()
-        self.bing_api_endpoint = self.config.get_bing_api_endpoint()
+        # self.config = Config()
```
Comment (Tineshnehete) [inline src/browser/search.py:13]: With config class
```suggestion
        self.config = Config()
```

**Comment by nityavp** [comment]
how to use this in devika?

**Comment by Vishnuxx** [comment]
@Tineshnehete Thanks once again for reviewing. I've made these changes. Now you can set the default search engine in config.toml file.

**Comment by Vishnuxx** [comment]
> how to use this in devika?

Hii @nityavp , you can use this feature right after the moderators merges this PR.

**Comment by ARajgor** [comment]
hey already these functionalities are been merged into the main so can we close this.

**Pull request closed without merging**